qiskit/extensions/quantum_initializer/state_preparation.py

```diff
@@ -0,0 +1,462 @@
+            final_t = 0
+        else:
+            theta = float(2 * np.arccos(mag_a / final_r))
+            a_arg = np.angle(a_complex)
+            b_arg = np.angle(b_complex)
+            final_t = a_arg + b_arg
+            phi = b_arg - a_arg
+
+        return final_r * np.exp(1.0j * final_t / 2), theta, phi
+
+    def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+        """
+        Return a recursive implementation of a multiplexor circuit,
+        where each instruction itself has a decomposition based on
+        smaller multiplexors.
+
+        The LSB is the multiplexor "data" and the other bits are multiplexor "select".
+
+        Args:
+            target_gate (Gate): Ry or Rz gate to apply to target qubit, multiplexed
+                over all other "select" qubits
+            list_of_angles (list[float]): list of rotation angles to apply Ry and Rz
+            last_cnot (bool): add the last cnot if last_cnot = True
+
+        Returns:
+            DAGCircuit: the circuit implementing the multiplexor's action
+        """
+        list_len = len(list_of_angles)
+        local_num_qubits = int(math.log2(list_len)) + 1
+
+        q = QuantumRegister(local_num_qubits)
+        circuit = QuantumCircuit(q, name="multiplex" + local_num_qubits.__str__())
+
+        lsb = q[0]
+        msb = q[local_num_qubits - 1]
+
+        # case of no multiplexing: base case for recursion
+        if local_num_qubits == 1:
+            circuit.append(target_gate(list_of_angles[0]), [q[0]])
+            return circuit
+
+        # calc angle weights, assuming recursion (that is the lower-level
+        # requested angles have been correctly implemented by recursion
+        angle_weight = np.kron([[0.5, 0.5], [0.5, -0.5]], np.identity(2 ** (local_num_qubits - 2)))
+
+        # calc the combo angles
+        list_of_angles = angle_weight.dot(np.array(list_of_angles)).tolist()
+
+        # recursive step on half the angles fulfilling the above assumption
+        multiplex_1 = self._multiplex(target_gate, list_of_angles[0 : (list_len // 2)], False)
+        circuit.append(multiplex_1.to_instruction(), q[0:-1])
+
+        # attach CNOT as follows, thereby flipping the LSB qubit
+        circuit.append(CXGate(), [msb, lsb])
+
+        # implement extra efficiency from the paper of cancelling adjacent
+        # CNOTs (by leaving out last CNOT and reversing (NOT inverting) the
+        # second lower-level multiplex)
+        multiplex_2 = self._multiplex(target_gate, list_of_angles[(list_len // 2) :], False)
+        if list_len > 1:
+            circuit.append(multiplex_2.to_instruction().reverse_ops(), q[0:-1])
+        else:
+            circuit.append(multiplex_2.to_instruction(), q[0:-1])
+
+        # attach a final CNOT
+        if last_cnot:
+            circuit.append(CXGate(), [msb, lsb])
+
+        return circuit
+
+
+def prepare_state(self, state, qubits=None):
+    r"""Prpare qubits in a specific state.
```

Comment:
This docstring (and similar for `Initialize`) could probably do with referencing each other, and explaining the difference between them.